pdns/packethandler.cc

```diff
@@ -1526,12 +1527,17 @@ std::unique_ptr<DNSPacket> PacketHandler::doQuestion(DNSPacket& p)
     }
     DLOG(g_log<<Logger::Error<<"We have authority, zone='"<<d_sd.qname<<"', id="<<d_sd.domain_id<<endl);
 
+    if (retargetcount == 0) {
+      r->qdomainzone = d_sd.qname;
+    } else if (!d_doResolveAcrossZones && r->qdomainzone != d_sd.qname) {
+      // We are following a retarget outside the initial zone. Config asked us not to do that.
+      goto sendit;  // NOLINT(cppcoreguidelines-avoid-goto)
```

Comment:
It is somewhat inefficient to bail out here, after most of the work is done (getAuth()).
An additional check if the target is part of the zone before getAuth() will save resources in most cases (all except for subdomains).

Reply:
Added a check there.